### PR TITLE
UI: Add obs_frontend_close_main_window()

### PR DESCRIPTION
**Pull request by BenJuan26**
### Description

Adds the ability to close the main window from the frontend API. The behaviour is the same as clicking the Exit button in OBS Studio.

### Motivation and Context

This is useful for scripting, where OBS is mainly run from the system tray and controlled through the script, i.e. through obs-websocket. Currently, the only way to close OBS from a script is to have the OS terminate the process. This is obviously not ideal: the program doesn't get the chance to properly free up resources, and in the case of Windows, the icon hangs around the system tray until it is moused-over.

Link to feature suggestion: https://ideas.obsproject.com/posts/2225/allow-clean-shutdown-via-api

### How Has This Been Tested?

OS: Windows
CPU: Intel i7 12700K
GPU: Gigabyte Nvidia GeForce RTX 3070 Ti 8 GB
RAM: 32GB

For the sake of simplicity, I hijacked the `StartStream` handler of obs-websocket to instead call `obs_frontend_close_main_window()`, then used OBS-web to trigger the handler. The OBS Studio window closed, and the log indicated that the shutdown process completed nominally, just as it would have if the Exit button was clicked.

### Types of changes

- New feature (non-breaking change which adds functionality)

- Documentation (a change to documentation pages)

### Checklist:

- [x] My code has been run through [clang-format](https://github.com/obsproject/obs-studio/blob/master/.clang-format).
- [x] I have read the [**contributing** document](https://github.com/obsproject/obs-studio/blob/master/CONTRIBUTING.rst).
- [x] My code is not on the master branch.
- [x] The code has been tested.
- [x] All commit messages are properly formatted and commits squashed where appropriate.
- [x] I have included updates to all appropriate documentation.

### CODE REVIEW

**Comment by PatTheMav** [comment]
Is this supposed to end the application? Because on macOS an application is not automatically terminated just because it has no active windows - that’s an opt-in behaviour. It could be that Qt behaves this way in our case, but that needs to be verified.

**Comment by BenJuan26** [comment]
For the use case that I'm targeting, yes, the intention is to end the application.

I just tested the release version of OBS on macOS, and indeed, clicking the Exit button and closing the window both end the application.

**Comment by derrod** [comment]
> Is this supposed to end the application? Because on macOS an application is not automatically terminated just because it has no active windows - that’s an opt-in behaviour. It could be that Qt behaves this way in our case, but that needs to be verified.

Our main window's close event handler does some cleanup and ultimately calls `qApplication::quit()`

https://github.com/obsproject/obs-studio/blob/c289efd84a42887feba89c3579d39a9c264b749e/UI/window-basic-main.cpp#L5042

File: docs/sphinx/reference-frontend-api.rst [hunk 1]
```diff
@@ -242,6 +242,12 @@ Functions
 
 ---------------------------------------
 
+.. function:: void obs_frontend_close_main_window(void)
+
+   Closes the main window.
```
Comment (CodeYan01) [inline docs/sphinx/reference-frontend-api.rst:247]: Perhaps some clarification here is in order. Like, would it prompt the user if there are active outputs, or ignore outputs.

Comment (BenJuan26) [reply docs/sphinx/reference-frontend-api.rst]: I added a line about active outputs. I didn't want to make things too wordy. I could have stated that behaviour is the same as the Exit button, but I think it's clearer to just state the behaviour.

Comment (CodeYan01) [reply docs/sphinx/reference-frontend-api.rst]: Currently looks good to me. However, I would prefer:
> Closes the main window**, as if the user clicked the Exit button**. If there are active outputs, the user will be prompted to confirm before closing.

Which may or may not be necessary, and I'll leave that opinion to you and the others.

Comment (BenJuan26) [reply docs/sphinx/reference-frontend-api.rst]: I'll leave it unresolved for now, so others can weigh in.

**Review by CodeYan01 — COMMENTED** [review]
LGTM

**Comment by Warchamp7** [comment]
I'd like us to include a log line indicating when something has closed OBS this way.

**Comment by asoulliereHT** [comment]
> I'd like us to include a log line indicating when something has closed OBS this way.
> 
> How does this interact with things that prevent closing the window such as running outputs?

"Closes the main window. If there are active outputs, the user will be prompted to confirm before closing."

**Comment by BenJuan26** [comment]
@Warchamp7 Log line added.

> How does this interact with things that prevent closing the window such as running outputs?

This was answered in the next comment. Let me know if you have any more questions/issues.

**Review by Warchamp7 — APPROVED** [review]
Fine from my end with the log line added

**Comment by GeneralSpace** [comment]
@BenJuan26 @Warchamp7 this would actually really help with a project I'm working on. Any chance it could be considered for inclusion in 31.1 or another minor version soon?

**Comment by BenJuan26** [comment]
@Warchamp7 @CodeYan01 This very simple change is nearly two years old and many parties have expressed interest in using it. Any idea whether there are plans to merge it anytime soon? Thanks.

**Comment by RayneYoruka** [comment]
It seems I'm landing here too looking for a way to gracefully shutdown obs. I don't think taskill is the best way to do it. 

The log ends up being incomplete as it counts as a forced exit.

**Comment by asoulliereHT** [comment]
I'm commenting here to bump.  I'm seeing wide support for this feature and no concerns on code.  🤷

**Comment by RayneYoruka** [comment]
I'm returning back with a hackery solution yet this seems to be the best! 
NirCMD can gracefully exit obs without any issues, just make a bat file and automate in your favourite software and have it have "nircmd.exe closeprocess obs64.exe"

It exists as if you used the X button to close!

**Comment by BenJuan26** [comment]
Closing as unnecessary. OBS now handles termination signals correctly.
